### DISCUSSION TITLE
Oven media hosted in azure behind Application Gateway / TURN / STUN / WebRTC

### DISCUSSION THREAD

**Discussion by mvximvs** [OvenMediaLabs/OvenMediaEngine#1547] [Help] [answered]
Hi Community <3 
I just wanted to ask for help related with the mix Azure + Application Gateway + Virtual Machine (private) + Oven Media + WebRTC
We have been using Oven Media since 3 years right now and first of all thanks for this amazing product guys, Oven Media rocks  !

We have been forced to move to a more secured environment in which Virtual Machines are private so we need to use an Application Gateway (azure) in front of every machine with TLS termination at the backend level.  The video stream arrives from an internal source so  no problem to push using the RTMP protocol. But now we are facing a problem with WebRTC in order to visualize the stream.

We have configured LLHLS on this scenario without problems, but we wanted to use WebRTC as the official technology to visualize our streams, and I'm currently facing some problems and I need guidance and clarification of concepts 🙏.

For those not familiar with the service [application gateway](https://learn.microsoft.com/en-us/azure/application-gateway/overview), it is a service that uses the application layer (7) of the OSI model so only http / ws , https / wss are allowed. 
In my case web socket secure works well (I can see the requests on logs) but my problem comes with TURN STUN servers.

First of all the cybersecurity department refuses to allow the **_stun.l.google.com:19302_** STUN server so on TCP relay I'm trying to set the IP of my application gateway.... and same thing in the IceCandidate node... but things are not working.... 
I was using fiddler in order to diagnostic the network and I see "tunnel to"   the "gateway_ip:3478" ... Do I need to allow this in my application gateway as well? SInce my application gateway only allows protocols + ip + ports... I'm struggling to authorize the turn server since I don't know what protocol is... 

I don't event know if what I'm trying to do using this application gateway it will be possible... Sorry if the description is not clear... 
Don't hesitate to ask me for more details or to re-explain this.

Thanks for your patient. 
Max

**Comment by basisbit** [comment]
there is absolutely no need to use google's STUN server, as long as you know the public IPv4 address over which the OME server is reachable.
However, the WebRTC signalling will not work unless you set the gateway to not break up the dTLS UDP connection, otherwise the fingerprint check for the dTLS will not pass and the client will not connect successfully. An external TURN server will not be of any use in such a use case.

In any case, running a low latency video streaming service with many Terabytes of outbound traffic is not something you want to do using AWS or Azure or similar very expensive per traffic environments. They are absolutely not optimized for this kind of usage and the application gateway does not add any security at all.

If you don't need low latency, change to using LLHLS or HLS and use your "application gateway" as a HTTP reverse proxy in front of it. If you need reliably and low latency, talk to your senior network engineers and make them understand what kind of service you are operating.

**Comment by getroot** [comment, marked as answer]
WebRTC transfers media data over RTP, ICE (TURN, STUN) protocols. Based on your description, I don't expect ICE Connect to work.
